src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/histogram.js

```diff
@@ -0,0 +1,533 @@
+   "data" : [
+     "values" : [0.1, 0.2, 0.1, 0.15, 0.7],
+   ]
+ }
+ */
+class Histogram extends Visualization {
+    static inputType = "Any"
+    static label     = "Histogram (JS)"
+
+    onDataReceived(data) {
+        const parsedData = JSON.parse(data);
+
+        while (this.dom.firstChild) {
+            this.dom.removeChild(this.dom.lastChild);
+        }
+        const isUpdate = parsedData.update === "diff";
+
+        this.updateData(parsedData, isUpdate)
+        this.updateHistogram(parsedData)
+    }
+
+    /**
+     * Update the internal data and plot settings with the ones from the new incoming data.
+     * If no new settings/data have been provided the old ones will be kept.
+     */
+    updateData(data, isUpdate) {
+        if (isUpdate) {
+            this.axisSpec    = data.axis         || this.axisSpec;
+            this.focus       = data.focus        || this.focus;
+            this.dataValues  = data.data.values  || this.data;
+            this.bins        = data.bins         || this.bins;
+        } else {
+            this.axisSpec    = data.axis;
+            this.focus       = data.focus;
+            this.dataValues  = data.data.values;
+            this.bins        = data.bins;
+        }
+    }
+
+    /**
+     * Return the focus area of the histogram. If none is set, returns undefined.
+     */
+    getFocus() {
+        return this.focus
+    }
+
+    /**
+     * Return specification for the axis. This includes scales (logarithmic/linear) and labels.
+     */
+    getAxisSpec() {
+        return this.axisSpec || {x:{scale:LINEAR_SCALE},y:{scale:LINEAR_SCALE}}
+    }
+
+    /**
+     * Return vales to plot.
+     */
+    getData() {
+        return this.dataValues || {};
+    }
+
+    /**
+     * Return the number of bins to use for the histogram..
+     */
+    getBinCount() {
+        return this.bins || DEFAULT_NUMBER_OF_BINS
+    }
+
+    /**
+     * Return the layout measurements for the plot. This includes the outer dimensions of the
+     * drawing area as well as the inner dimensions of the plotting area and the margins.
+     */
+    getCanvasDimensions() {
+        const width  = this.dom.getAttributeNS(null,"width");
+        let height   = this.dom.getAttributeNS(null,"height");
+        height       = height - BUTTON_HEIGHT;
+        const margin = this.getMargins(this.getAxisSpec());
+        return {
+            inner : {
+                width  : width - margin.left - margin.right,
+                height : height - margin.top - margin.bottom
+            },
+            outer : {
+                width  : width,
+                height : height,
+            },
+            margin : margin,
+        };
+    }
+
+    /**
+     * Update the histogram with the current data and settings.
+     */
+    updateHistogram() {
+        const canvas  = this.getCanvasDimensions();
+        const divElem = this.createDivElem(canvas.outer.width,canvas.outer.height);
+        this.dom.appendChild(divElem);
```

Comment:
Why during the update function we are modyfying the DOM? Sounds like a serious performance issue for me, but I might be wrong. Could you tell more about it ?

**EDIT** Oh, yo uare re-drawing the histogram from scratch here, right? If so, where is the function shich drops the old div appended to dom?